Update request-response.md (#35289)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/middleware/request-response.md

* `TextWriter`
* `HttpResponse.WriteAsync`

Streams aren't being removed from the framework. Streams continue to be used throughout .NET:

* Many stream types don't have pipe equivalents, such as `FileStreams` and `ResponseCompression`.

> * It's straightforward adding compression to the stream.

Author: ```suggestion
* It's straightforward adding compression to a stream.
```

* `TextWriter`
* `HttpResponse.WriteAsync`

> Streams aren't being removed from the framework. Streams continue to be used throughout .NET, and many stream types don't have pipe equivalents, such as `FileStreams` and `ResponseCompression`, and straightforward adding compression to the stream.

Reviewer: I understand that it is not possible to write a custom compression middleware with PipeWriter. Is that correct?

Author: > @Rick-Anderson could you also cover what I mentioned in [#35288 (comment)](https://github.com/dotnet/AspNetCore.Docs/issues/35288#issuecomment-2822055748)

That's what I tried to do here (line 28 change).

Author: > I understand that it is not possible to write a custom compression middleware with PipeWriter. Is that correct?

@gfoidl @martincostello  ???

Reviewer 2: > not possible to write a custom compression middleware with PipeWriter

It's possible, but very awkward compared to Streams. W/ Streams it's quite easy, as the decorator pattern can be used to wrap a stream, which wraps a streams, etc.

W/ PipeWriter more custom code would be needed. E.g. an intermediate Pipeline that reads what the user's PipeWriter writes (and flushes), then perfom the compression and write (and flush) it the response body's PipeReader.
Also for compression .NET provides no easy to use infrastructure for this use case.

So as said it's possible, but w/ quite an effort. So I see no reason why one should use PipeWriter for this, when it's quite easy to accomplish with Streams.

Note: The Stream in the response is quite a simple adapter for the PipeWriter (at least in Kestrel), so there shouldn't be any perf-difference, especially when the data is sent over the network. So use either `Body` or `BodyWriter` depending on the use case and on how it can be done more easily.

Reviewer: > The Stream in the response is quite a simple adapter for the PipeWriter

I think this is the piece that I was missing. The pipe and stream are different APIs for the same channel of data? For example, if my endpoint middleware writes to pipe and my compression middleware wraps the stream, my data will be compressed?

Could we explicit this part in this document 🙏

Reviewer 2: > The pipe and stream are different APIs for the same channel of data?

Yep.
Underlying is the network socket, from which data is written / read. In oder to make these operations easier to use and abstracted away (because instead of network socket it could also be named pipes, unix domain sockets, etc. -- effort was done in ASP.NET Core with so called "project bedrock") there's pipes and streams, which allow "unified" access higher up in the stack.

> if my endpoint middleware writes to pipe and my compression middleware wraps the stream, my data will be compressed?

W/o knowing your code I'm not able to say yes or no.
Interleaving PipeWriter and BodyStream is dangerous, as none of these is thread-safe. They're not intended to be used concurrent or mixed in any way.
So it really depends when you write / flush the pipe and what you'll do with the stream.

Reviewer: Let's say my endpoint handler is just writing a protobuf object to the pipe
```csharp
MyProtoObject.WriteTo(HttpContext.Response.BodyWriter)
```
and I use the standard response compression middleware
```csharp
app.UseResponseCompression();
```

I'm expecting the compression middleware to decorate the stream and not touch the pipe. I'm wondering if my response will be compressed.

> Streams aren't being removed from the framework. Streams continue to be used throughout .NET, and many stream types don't have pipe equivalents, such as `FileStreams` and `ResponseCompression`, and straightforward adding compression to the stream.

Reviewer 2: ```suggestion
Streams aren't being removed from the framework. Streams continue to be used throughout .NET, and many stream types don't have pipe equivalents, such as `FileStreams` and `ResponseCompression`, and it's straightforward adding compression to the stream.
```

> Streams aren't being removed from the framework. Streams continue to be used throughout .NET, and many stream types don't have pipe equivalents, such as `FileStreams` and `ResponseCompression`, and it's straightforward adding compression to the stream.

Author: ```suggestion
Streams aren't being removed from the framework. Streams continue to be used throughout .NET:

* Many stream types don't have pipe equivalents, such as `FileStreams` and `ResponseCompression`.
* It's straightforward adding compression to the stream.
```

## Pipelines

When writing directly to `HttpResponse.BodyWriter`, call `PipeWriter.FlushAsync()` manually to ensure the data is flushed to the underlying response response body. Here's why:

* `HttpResponse.BodyWriter` is a `PipeWriter` that buffers data until a flush operation is triggered.

> * Calling `FlushAsync` writes the buffered data to the underlying network operation

Author: ```suggestion
* Calling `FlushAsync` writes the buffered data to the underlying network operation

```

## Pipelines

> When writing directly to `HttpResponse.BodyWriter`, call `PipeWriter.FlushAsync()` manually to ensure the data is flushed to the underlying response stream. Here's why:

Reviewer 2: ```suggestion
When writing directly to `HttpResponse.BodyWriter`, call `PipeWriter.FlushAsync()` manually to ensure the data is flushed to the underlying response data structure. Here's why:
```
Actually it's not a `Stream`. For Kestrel the data is written to socket send more or less directly.
Thus the second bullet below is also a bit misleading.

Reviewer 3: `response body`?

> When writing directly to `HttpResponse.BodyWriter`, call `PipeWriter.FlushAsync()` manually to ensure the data is flushed to the underlying response response body. Here's why:

Author: ```suggestion
When writing directly to `HttpResponse.BodyWriter`, call `PipeWriter.FlushAsync` manually to ensure the data is flushed to the underlying response response body. Here's why:
```

* `HttpResponse.BodyWriter` is a `PipeWriter` that buffers data until a flush operation is triggered.
* Calling `FlushAsync` writes the buffered data to the underlying response body

> 

Author: ```suggestion
It's up to the developer to decide when to call `FlushAsync`, balancing factors such as buffer size, network write overhead, and whether the data should be sent in discrete chunks. For more information, see [System.IO.Pipelines in .NET](/dotnet/standard/io/pipelines).
```

> * Calling `FlushAsync` writes the buffered data to the underlying network operation

Author: ```suggestion
* Calling `FlushAsync` writes the buffered data to the underlying response body
```

> * Calling `FlushAsync` writes the buffered data to the underlying Stream (`HttpResponse.Body`).

Reviewer: Should we specify how frequently the FlushAsync should be called? Once? Once per Advance? Every X bytes written?

Author: > Should we specify how frequently the FlushAsync should be called? Once? Once per Advance? Every X bytes advanced?

@gfoidl @martincostello ???

Reviewer 2: > My guess is it's hard to give a simple answer.

True. A generic rule when to call `Flush` is hard to impossible to give.

Very general:
* call `Flush` to make the data available to the `PipeReader` (here: the response body), so that the response can be transmitted (chunked) over the wire to the client
* call `Flush` when someting is written that the `PipeReader` should have access to (before completing the pipe)

But I don't know how to defer a rule out of the points.
Honestly, I'd just omit something like this, as at the end of the article there's a link to _System.IO.Pipelines in .NET_ which has further infos.

Reviewer: Could we say something like "It's up to the developer to decide when to call FlushAsync, balancing factors such as buffer size, network write overhead, and whether the data should be sent in discrete chunks."

> * Calling `FlushAsync` writes the buffered data to the underlying Stream (`HttpResponse.Body`).

Reviewer 2: ```suggestion
* Calling `FlushAsync` writes the buffered data to the underlying network operation
```
?

> * 

Reviewer 2: This bullet should be a blank line instead?
```suggestion

```

> It's up to the developer to decide when to call `FlushAsync`, balancing factors such as buffer size, network write overhead, and whether the data should be sent in discrete chunks. For more information, see [System.IO.Pipelines in .NET](/dotnet/standard/io/pipelines).

Author: ```suggestion

It's up to the developer to decide when to call `FlushAsync`, balancing factors such as buffer size, network write overhead, and whether the data should be sent in discrete chunks. For more information, see [System.IO.Pipelines in .NET](/dotnet/standard/io/pipelines).
```

## Pipelines

When writing directly to `HttpResponse.BodyWriter`, call `PipeWriter.FlushAsync` manually to ensure the data is flushed to the underlying response response body. Here's why:

* `HttpResponse.BodyWriter` is a `PipeWriter` that buffers data until a flush operation is triggered.

> * Calling `FlushAsync` writes the buffered data to the underlying response body

Reviewer 3: ```suggestion
* Calling `FlushAsync` writes the buffered data to the underlying response body.
```

## Pipelines

> The following example shows how the preceding stream scenario can be handled using a [PipeReader](/dotnet/standard/io/pipelines#pipe):

Author: ```suggestion
The following example shows how the preceding stream scenario can be handled using a [PipeReader](/dotnet/standard/io/pipelines#pipe):

```

* Encoded strings are directly added to the list of returned strings.
* Other than the `ToArray` call, and the memory used by the string, string creation is allocation free.

When writing directly to `HttpResponse.BodyWriter`, call `PipeWriter.FlushAsync` manually to ensure the data is flushed to the underlying response response body. Here's why:

* `HttpResponse.BodyWriter` is a `PipeWriter` that buffers data until a flush operation is triggered.
* Calling `FlushAsync` writes the buffered data to the underlying response body

> It's up to the developer to decide when to call `FlushAsync`, balancing factors such as buffer size, network write overhead, and whether the data should be sent in discrete chunks. For more information, see [System.IO.Pipelines in .NET](/dotnet/standard/io/pipelines).

Author: ```suggestion
It's up to the developer to decide when to call `FlushAsync`, balancing factors such as buffer size, network write overhead, and whether the data should be sent in discrete chunks. For more information, see [System.IO.Pipelines in .NET](/dotnet/standard/io/pipelines).

```

* Encoded strings are directly added to the list of returned strings.
* Other than the `ToArray` call, and the memory used by the string, string creation is allocation free.

When writing directly to `HttpResponse.BodyWriter`, call `PipeWriter.FlushAsync` manually to ensure the data is flushed to the underlying response response body. Here's why:

* `HttpResponse.BodyWriter` is a `PipeWriter` that buffers data until a flush operation is triggered.

> * Calling `FlushAsync` writes the buffered data to the underlying response body

Author: ```suggestion
* Calling `FlushAsync` writes the buffered data to the underlying response body.
```
